mkosi/util.py

```diff
@@ -221,49 +226,36 @@ def flatten(lists: Iterable[Iterable[T]]) -> list[T]:
     return list(itertools.chain.from_iterable(lists))
 
 
-@dataclasses.dataclass
 class InvokingUser:
-    _pw: Optional[pwd.struct_passwd] = None
+    @staticmethod
+    def _uid_from_env() -> Optional[str]:
+        return os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID")
 
     @classmethod
-    def for_uid(cls, uid: int) -> "InvokingUser":
-        return cls(pwd.getpwuid(uid))
-
-    @property
-    def uid(self) -> int:
-        if self._pw is not None:
-            return self._pw.pw_uid
-        return os.getuid()
-
-    @property
-    def gid(self) -> int:
-        if self._pw is not None:
-            return self._pw.pw_gid
-        return os.getgid()
-
-    @property
-    def name(self) -> str:
-        if self._pw is not None:
-            return self._pw.pw_name
-        return getpass.getuser()
+    def uid(cls) -> int:
+        return int(cls._uid_from_env() or os.getuid())
 
-    @property
-    def home(self) -> Path:
-        if self._pw is not None:
-            return Path(self._pw.pw_dir)
-        return Path.home()
+    @classmethod
+    def uid_gid(cls) -> tuple[int, int]:
+        if uid := cls._uid_from_env:
+            gid = int(os.getenv("SUDO_GID") or pwd.getpwuid(uid).pw_gid)
+            return uid, gid
+        return os.getuid(), os.getgid()
 
-    def is_running_user(self) -> bool:
-        if self._pw is not None:
-            return self._pw.pw_uid == os.getuid()
-        return True
+    @classmethod
+    def name(cls) -> str:
+        if uid := cls._uid_from_env():
+            return pwd.getpwuid(uid).pw_name
+        return getpass.getuser()
 
+    @classmethod
+    def home(cls) -> Path:
+        home = os.getenv("SUDO_HOME") or os.getenv("HOME") or pwd.getpwuid(cls.uid()).pw_dir
```

Comment:
This is wrong, it returns `Path("/root")` with sudo and pkexec. `$SUDO_HOME` is not a real environment variable set by sudo, but a fake one that was set by `--extra-search-path` when it was first introduced and now replaced by expanding `~`.